File: app/src/main/java/me/devsaki/hentoid/notification/download/DownloadProgressNotification.java
package me.devsaki.hentoid.notification.download;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.core.content.ContextCompat;

import java.util.Locale;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.activities.QueueActivity;
import me.devsaki.hentoid.util.notification.Notification;

public class DownloadProgressNotification implements Notification {

    private final String title;

    private final int progress;

    private final int max;

    public DownloadProgressNotification(String title, int progress, int max) {
        this.title = title;
        this.progress = progress;
        this.max = max;
    }

    @NonNull
    @Override
    public android.app.Notification onCreateNotification(Context context) {
        return new NotificationCompat.Builder(context, DownloadNotificationChannel.ID)
                .setSmallIcon(R.drawable.ic_stat_hentoid)
                .setContentTitle(context.getString(R.string.downloading))
                .setContentText(title)
                .setContentInfo(getProgressString())
                .setProgress(max, progress, false)
                .setColor(ContextCompat.getColor(context, R.color.secondary))
                .setContentIntent(getDefaultIntent(context))
                .setLocalOnly(true)
                .setOngoing(true)
                .setOnlyAlertOnce(true)
                .build();
    }

    private PendingIntent getDefaultIntent(Context context) {
        Intent resultIntent = new Intent(context, QueueActivity.class);
        resultIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_SINGLE_TOP);

        return PendingIntent.getActivity(context, 0, resultIntent, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    private String getProgressString() {
        return String.format(Locale.US, " %.2f%%", progress * 100.0 / max);
    }
}

File: app/src/main/java/me/devsaki/hentoid/notification/download/DownloadProgressNotification.kt
package me.devsaki.hentoid.notification.download

import android.app.PendingIntent
import android.content.Context
import androidx.core.app.NotificationCompat
import me.devsaki.hentoid.R
import me.devsaki.hentoid.activities.QueueActivity
import me.devsaki.hentoid.receiver.DownloadNotificationCancelReceiver
import me.devsaki.hentoid.receiver.DownloadNotificationPauseReceiver
import me.devsaki.hentoid.util.getThemedColor
import me.devsaki.hentoid.util.notification.BaseNotification
import java.util.Locale

class DownloadProgressNotification() : BaseNotification() {
    var title = ""
    var progress = 0
    var max = 0
    var sizeDownloadedMB = 0
    var estimateBookSizeMB = 0
    var avgSpeedKbps = 0

    lateinit var builder: NotificationCompat.Builder

    override fun onCreateNotification(context: Context): android.app.Notification {
        val progressString: String = " %.2f%%".format(Locale.US, progress * 100.0 / max)
        val total = if (estimateBookSizeMB > -1) "/$estimateBookSizeMB" else ""
        val message =
            context.getString(R.string.download_notif_speed, sizeDownloadedMB, total, avgSpeedKbps)

        if (!this::builder.isInitialized) {
            builder = NotificationCompat.Builder(context, ID)
                .setSmallIcon(R.drawable.ic_app)
                .setColor(context.getThemedColor(R.color.secondary_light))
                .setContentIntent(getDefaultIntent(context))
                .addAction(
                    R.drawable.ic_action_pause,
                    context.getString(R.string.pause),
                    getPauseIntent(context)
                )
                .addAction(
                    R.drawable.ic_cancel,
                    context.getString(R.string.cancel),
                    getCancelIntent(context)
                )
                .setLocalOnly(true)
                .setOngoing(true)
                .setOnlyAlertOnce(true)
        }

        builder.apply {
            setContentTitle(title)
            setContentText(message)
            setContentInfo(progressString)
            setProgress(max, progress, 0 == max)
        }

        return builder.build()
    }

    private fun getDefaultIntent(context: Context): PendingIntent {
        return getPendingIntentForActivity(context, QueueActivity::class.java)
    }

    private fun getPauseIntent(context: Context): PendingIntent {
        return getPendingIntentForAction(context, DownloadNotificationPauseReceiver::class.java)
    }

    private fun getCancelIntent(context: Context): PendingIntent {
        return getPendingIntentForAction(context, DownloadNotificationCancelReceiver::class.java)
    }
}
